src/jvm/main/org/jetbrains/kotlinx/lincheck/runner/ParallelThreadsRunner.kt

```diff
@@ -308,35 +308,78 @@ internal open class ParallelThreadsRunner(
+            return RunnerTimeoutInvocationResult(threadDump, collectExecutionResults())
         } catch (e: ExecutionException) {
-            return UnexpectedExceptionInvocationResult(e.cause!!)
+            return UnexpectedExceptionInvocationResult(e.cause!!, collectExecutionResults())
         } finally {
             resetState()
         }
     }
 
+    /**
+     * This method is called when we have some execution result other than [CompletedInvocationResult].
+     */
+    fun collectExecutionResults(): ExecutionResult {
+        return collectExecutionResults(null, null, null)
+    }
+
+    private fun collectExecutionResults(
+        afterInitStateRepresentation: String?,
+        afterParallelStateRepresentation: String?,
+        afterPostStateRepresentation: String?
+    ): ExecutionResult {
+        val initResults = initialPartExecution?.results?.let { collectExecutionResults(it, INIT) }.orEmpty()
+        val parallelResultsWithClock = parallelPartExecutions.map { execution ->
+            collectExecutionResults(execution.results, PARALLEL).zip(execution.clocks).map {
+                ResultWithClock(it.first, HBClock(it.second.clone()))
+            }
+        }
+        val postResults = postPartExecution?.results?.let { collectExecutionResults(it, POST) }.orEmpty()
+
+        return ExecutionResult(
+            initResults = initResults,
+            parallelResultsWithClock = parallelResultsWithClock,
+            postResults = postResults,
+            afterInitStateRepresentation = afterInitStateRepresentation,
+            afterParallelStateRepresentation = afterParallelStateRepresentation,
+            afterPostStateRepresentation = afterPostStateRepresentation
+        )
+    }
+
+    /**
+     * Composes a list of execution results from a raw array representation.
+     * When execution is finished in a normal way we have all non-nullable [TestThreadExecution.results].
+     * But when the execution has hung, results for some actors may be not set.
+     * This method checks if we have some unset results and if so, it replaces the first one as [NotFinished],
+     * and the rest as [NotStarted].
+     *
+     * @param executionPart indicates what execution part this result correspond to
+     */
+    private fun collectExecutionResults(rawResults: Array<Result?>, executionPart: ExecutionPart): List<Result> {
+        // If execution ended in the init part, then all actors after it was not started.
+        if (currentExecutionPart == INIT && executionPart != INIT) return List(rawResults.size) { NotStarted }
+        // If execution ended in the parallel part, then all actors after it was not started.
+        if (currentExecutionPart == PARALLEL && executionPart == POST) return List(rawResults.size) { NotStarted }
+        // Find the first unset result - it means this actor was started but not finished.
+        val nullFirstNotSetResultIndex = rawResults.indexOfFirst { it == null }
```

Comment:
I think this logic is specific to results representation, not the model itself.

When we print the results to the console, we  want to remove all not-started actors and display only the first non-finished (hung) actor.
In other cases, we might want to handle these results differently.
So it is better to move this logic to the `Reporter.kt`.